### DISCUSSION TITLE
example how to implement new datasource that can be discovered with unchanged renovate installation

### DISCUSSION THREAD

**Discussion by ghost** [renovatebot/renovate#22975] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

latest

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

Hello renovate community :)

I want to implement new datasource (based at [artifactory](https://github.com/renovatebot/renovate/tree/main/lib/modules/datasource/artifactory) but using search api and later will try one for Debian repositories) that can be called via unchanged renovate image.

Is there a known example for own `external` datasource registring and bootstraping original renovate.js ?

BR
Andre

### What you tried so far.

I 've searched a lot and found some ideas and one example that i do not really understand:  https://github.com/containerbase/internal-tools/blob/de9d26ed272e3cc0b8c5af8ba47ff59eccc47e56/src/utils/datasource/index.ts

I've copied the artifactory datasource and renamed it to `artifactory-search` and try to bootstrap with calling register own datasource and calling the original renovate:

renovate.ts:
```ts
import * as datasource from './datasource/datasource';
import * as renovate from 'renovate/dist/renovate';
import { logger } from 'renovate/dist/logger';

export function register(): void {
  logger.debug('register renovate extensions');
  datasource.register();
}

register();

renovate

```

datasource.ts:
```ts
import { logger } from 'renovate/dist/logger';
import { getDatasources } from 'renovate/dist/modules/datasource';
import { ArtifactorySearchDatasource } from './artifactory-search';

export function register(): void {
    logger.debug('register datasource extensions');
    const api = getDatasources();
    api.set(ArtifactorySearchDatasource.id, new ArtifactorySearchDatasource());
}
```

I see that registering happens and new datasource in the `api` map, but during processing (via reg-ex manager) the datasource name is known but `getReleases()` is not called (breakpoint not hit) and only reports, see below, that related packages info cannot be fetched. 

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Found package lookup warnings (repository=oie-jupiter/product/platform/jupiter-platform-image)
       "warnings": [
         "Failed to look up artifactory-search package librsc-dev",
         "Failed to look up artifactory-search package rsc-services"
       ],
       "warningFiles": [
         ".scripts/development/libs/requirements.txt",
         ".scripts/runtime/libs/requirements.txt"
       ]
```

</details>

**Comment by rarkins** [comment]
It's not possible without changing the Renovate image

**Reply by ghost** [reply]
thanks, sorry i should formulate it better-  we can change the image itself, we install own tooling etc., but do not want to touch the installed renovate (latest version).

**Reply by rarkins** [reply]
I think it would be best if you integrate it from source to validate your implementation first. Then you will see which files need updating.

**Reply by ghost** [reply]
e.g. i've found an example with touching the renovate installation: https://repo1.dso.mil/dsop/container-hardening-tools/renovate/renovate/-/blob/development/Dockerfile#L48

and asking if it is possible without, because found this tip here: https://github.com/renovatebot/renovate/issues/4674#issuecomment-669021296

> You need your own renovate.js file where you add your customizations, before require the main renovate js.


but may be misunderstood...

**Reply by rarkins** [reply]
That's a pretty old comment, I think we changed the approach since then. So you probably understood it right but it's now outdated, I will hide that comment

**Reply by ghost** [reply]
Thanks. Now i tried to do it insource and just configured and debug original artifactory data source - same here i do not see why the lookup is not working and breakpoints to 'async getReleases({ packageName, registryUrl, })' are just not stopping - it only stopped once -  and now always the messages at the end:
```
EBUG: Failed to look up artifactory package librsc-dev (repository=oie-jupiter/product/platform/jupiter-platform-image, packageFile=.scripts/development/libs/requirements.txt, dependency=librsc-dev)
DEBUG: Failed to look up artifactory package rsc-services (repository=oie-jupiter/product/platform/jupiter-platform-image, packageFile=.scripts/runtime/libs/requirements.txt, dependency=rsc-services)
```

Is there a special behavior or how to debug or explanation why `getReleases()` is not called or just called once?

I have installed renovate locally...

**Reply by rarkins** [reply]
You may have a package cache bypassing lookups if you run the same repos repeatedly. You need to locate where the temporary directory is and delete it. For me it's `/tmp/renovate` but it's different per platform. Look for logs like this:

```
DEBUG: Using configured baseDir: /tmp/renovate
DEBUG: Using cacheDir: /tmp/renovate/cache
```

**Reply by ghost** [reply]
Thanks 👍 now it works, also with my first approach without touching renovate installation. I will test further and provide info how and integrate later in this repo...

**Reply by viceice** [reply]
but remember, it's not an official API. it can break any time in a patch release. even our own internal-tool action breaks regularly 🙃

**Reply by ghost** [reply]
Thanks, i'm aware that this is not an official API and will not complain if it breaks ;)
